### PR TITLE
Remove parameter from device.sample, make device.shots changeable and add an indicator for analytic mode

### PR DESCRIPTION
**Pull request by johannesjmeyer**
**Context:**
The user interface for sampling was suboptimal and caused significant complication in plugin implementation.

**Description of the Change:**
See #305:
* A new boolean parameter `analytic` is introduced for simulator plugins that indicates that analytic expectation values are to be calculated. This supersedes the old notation `shots=0`.
* The parameter `n` in `device.sample` is removed. The number of samples returned is always `device.shots`
* The parameter `device.shots` now defaults to 1 and can now be changed via `device.set_shots`

**Benefits:**
The user interface becomes clearer and the implementation of sampling in the plugins is significantly decluttered.

**Possible Drawbacks:**
Sample sizes can no more be specified directly in `dev.sample` but have to be changed at device level via `device.set_shots`.

**Related GitHub Issues:**
#305
_____
I was unable to compile the docs to check because the examples fail to compile. This is understandable as the plugin devices themselves have to be updated.

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -2,17 +2,30 @@
...
+  [#317](https://github.com/XanaduAI/pennylane/pull/317)
+
 ### Documentation
 
 ### Bug fixes
 
 ### Contributors
 
-This release contains contributions from (in alphabetical order):
+This release contains contributions from (in alphabetical order): Johannes Jakob Meyer
```
Comment (josh146) [inline .github/CHANGELOG.md]:
```suggestion
This release contains contributions from (in alphabetical order):

Johannes Jakob Meyer
```

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -2,17 +2,30 @@
...
 ### Breaking changes
 
+* The argument `n` specifying the number of samples in the method `Device.sample` was removed.
+  Instead, the method will always return `Device.shots` many samples. 
+  [#317](https://github.com/XanaduAI/pennylane/pull/317)
+
 ### Improvements
 
+* The number of shots / random samples used to estimate expectation values and variances, `Device.shots`,
+  can now be changed with the setter method `Device.set_shots`.
```
Comment (josh146) [inline .github/CHANGELOG.md]:
```suggestion
  can now be changed with the setter method `Device.set_shots()`.
```

File: .github/CHANGELOG.md [hunk 3]
```diff
@@ -2,17 +2,30 @@
 
 ### New features since last release
 
+* The plugins default.qubit and default.gaussian have a new initialization parameter
```
Comment (josh146) [inline .github/CHANGELOG.md]:
```suggestion
* The devices `default.qubit` and `default.gaussian` have a new initialization parameter
```

File: pennylane/_device.py [hunk 4]
```diff
@@ -175,6 +174,18 @@ def observables(self):
         """
         raise NotImplementedError
 
+    def set_shots(self, shots):
```
Comment (josh146) [inline pennylane/_device.py]: Should we also make `shots` a property, and internally use `self._shots`? This will avoid users attempting to change `shots` and causing internally consistency issues.

We could even go further, and use the `@shots.setter` decorator here, which would allow the user to use the syntax
```
dev.shots = 32
```
but still call the logic/validation in this method.

File: pennylane/_device.py [hunk 5]
```diff
@@ -175,6 +174,18 @@ def observables(self):
...
+        """Changes the number of shots.
+
+        Args:
+            shots (int): number of circuit evaluations/random samples used to estimate
+            expectation values of observables.
+        """
+        if shots < 1:
+            raise DeviceError("The specified number of shots needs to be at least 1. Got {}.".format(shots))
+
+        self.shots = shots
```
Comment (josh146) [inline pennylane/_device.py]:
```suggestion
        self.shots = int(shots)
```
We could also do input validation that shots is integral, but lets instead take advantage of duck typing :laughing:

File: pennylane/_device.py [hunk 6]
```diff
@@ -228,9 +239,7 @@ def execute(self, queue, observables, parameters={}):
                 elif obs.return_type is Variance:
                     results.append(self.var(obs.name, obs.wires, obs.parameters))
                 elif obs.return_type is Sample:
-                    if not hasattr(obs, "num_samples"):
-                        raise DeviceError("Number of samples not specified for observable {}".format(obs.name))
-                    results.append(np.array(self.sample(obs.name, obs.wires, obs.parameters, obs.num_samples)))
+                    results.append(np.array(self.sample(obs.name, obs.wires, obs.parameters)))
```
Comment (josh146) [inline pennylane/_device.py]: Much cleaner!

File: pennylane/_device.py [hunk 7]
```diff
@@ -422,8 +431,8 @@ def var(self, observable, wires, par):
         """
         raise NotImplementedError("Returning variances from QNodes not currently supported by {}".format(self.short_name))
 
-    def sample(self, observable, wires, par, n=None):
-        """Return a sample of an observable.
+    def sample(self, observable, wires, par):
+        """Return a sample of an observable with a size of ``device.shots``.
```
Comment (josh146) [inline pennylane/_device.py]:
```suggestion
        """Return a sample of an observable.
        
         The number of samples is determined by the value of ``device.shots``,
         and can be modified via...
```

Comment (josh146) [reply pennylane/_device.py]: Note: don't commit the above until the `...` is expanded!

File: pennylane/_device.py [hunk 8]
```diff
@@ -432,8 +441,6 @@ def sample(self, observable, wires, par, n=None):
             observable (str): name of the observable
             wires (Sequence[int]): subsystems the observable is to be measured on
             par (tuple): parameters for the observable
-            n (int): Number of samples that should be obtained. Defaults to the
-                number of shots given as a parameter to the corresponding Device.
```
Comment (josh146) [inline pennylane/_device.py]: :+1:

File: pennylane/measure.py [hunk 9]
```diff
@@ -148,40 +148,25 @@ def var(op):
...
         # delete operation from QNode queue
         QNode._current_context.queue.remove(op)
 
     # set return type to be a sample
     op.return_type = Sample
 
-    # attach the number of samples to the operation object
-    op.num_samples = n
-
     if QNode._current_context is not None:
```
Comment (josh146) [inline pennylane/measure.py]: Very happy with the simplification!

File: pennylane/plugins/default_gaussian.py [hunk 10]
```diff
@@ -701,10 +701,12 @@ class DefaultGaussian(Device):
 
     Args:
         wires (int): the number of modes to initialize the device in
-        shots (int): How many times should the circuit be evaluated (or sampled) to estimate
-            the expectation values. 0 yields the exact result.
+        shots (int): (default 1) How many times should the circuit be evaluated (or sampled) to estimate
```
Comment (josh146) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
        shots (int): How many times should the circuit be evaluated (or sampled) to estimate
```
The default value is automatically extracted from the signature by Sphinx, so no need to replicate it here :slightly_smiling_face:

File: pennylane/plugins/default_gaussian.py [hunk 11]
```diff
@@ -701,10 +701,12 @@ class DefaultGaussian(Device):
...
+            the expectation values.
         hbar (float): (default 2) the value of :math:`\hbar` in the commutation
             relation :math:`[\x,\p]=i\hbar`
+        analytic (bool): (default True) indicates if the device should calculate expectations
```
Comment (josh146) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
        analytic (bool): indicates if the device should calculate expectations
```

File: pennylane/plugins/default_gaussian.py [hunk 12]
```diff
@@ -823,7 +827,7 @@ def expval(self, observable, wires, par):
 
         ev, var = self._observable_map[observable](mu, cov, wires, par, self.num_wires, hbar=self.hbar)
 
-        if self.shots != 0:
+        if not self.analytic:
```
Comment (josh146) [inline pennylane/plugins/default_gaussian.py]: much better!

File: pennylane/plugins/default_qubit.py [hunk 13]
```diff
@@ -296,7 +296,9 @@ class DefaultQubit(Device):
     Args:
         wires (int): the number of modes to initialize the device in
         shots (int): How many times the circuit should be evaluated (or sampled) to estimate
-            the expectation values. A value of 0 yields the exact result.
+            the expectation values. Defaults to 1 if not specified.
+        analytic (bool): Indicates if the device should calculate expectations
+            and variances analytically. Defaults to True if not specified.
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]:
```suggestion
            and variances analytically
```

File: pennylane/plugins/default_qubit.py [hunk 14]
```diff
@@ -296,7 +296,9 @@ class DefaultQubit(Device):
     Args:
         wires (int): the number of modes to initialize the device in
         shots (int): How many times the circuit should be evaluated (or sampled) to estimate
-            the expectation values. A value of 0 yields the exact result.
+            the expectation values. Defaults to 1 if not specified.
+        analytic (bool): Indicates if the device should calculate expectations
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]:
```suggestion
        analytic (bool): indicates if the device should calculate expectations
```

File: pennylane/plugins/default_qubit.py [hunk 15]
```diff
@@ -296,7 +296,9 @@ class DefaultQubit(Device):
     Args:
         wires (int): the number of modes to initialize the device in
         shots (int): How many times the circuit should be evaluated (or sampled) to estimate
-            the expectation values. A value of 0 yields the exact result.
+            the expectation values. Defaults to 1 if not specified.
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]: There is a bit of a subtlety here.

* If `analytic=True`, ``shots`` is not used to calculate expval and var, but it does determine the number of samples returned.

* If `analytic = False`, ``shots`` _is_ used to calculate expval and var, and continues to determine the number of samples returned.

I think this could be made more clear in the docstring here (and in other relevant places)

File: tests/test_default_gaussian.py [hunk 16]
```diff
@@ -419,7 +419,7 @@ def test_apply_errors(self):
 
         with self.assertRaisesRegex(ValueError, "Invalid target subsystems provided in 'wires' argument"):
             p = U2
-            dev = DefaultGaussian(wires=4, shots=0, hbar=hbar)
+            dev = DefaultGaussian(wires=4, shots=1, hbar=hbar, analytic=True)
```
Comment (josh146) [inline tests/test_default_gaussian.py]: This test should have continued to pass without needing to add `analytic=True`, right?

File: tests/test_default_qubit.py [hunk 17]
```diff
@@ -545,6 +545,21 @@ def test_expval_warnings(self, qubit_device_1_wire):
...
+
+        @qml.qnode(dev)
+        def circuit():
+            return qml.expval(qml.PauliX(0))
+
+        expval = circuit()
+
+        # With 3 samples we are guaranteed to see a difference between
+        # an estimated variance an an analytically calculated one
+        assert expval != 0.0
```
Comment (josh146) [inline tests/test_default_qubit.py]: lol, nice

File: tests/test_measure.py [hunk 18]
```diff
@@ -304,60 +280,6 @@ def circuit():
...
-        @qml.qnode(dev)
-        def circuit():
-            return qml.sample(qml.PauliZ(0))
-
-        result = circuit()
-
-        # If all the dimensions are equal the result will end up to be a proper rectangular array
-        assert np.array_equal(result.shape, (n_shots,))
-
-    def test_sample_exception_device_context_missing(self):
```
Comment (co9olguy) [inline tests/test_measure.py]: How come this test was removed? I expected it to show up in `test_device.py` now, but didn't see it there

File: tests/test_default_qubit.py [hunk 19]
```diff
@@ -653,42 +671,12 @@ def test_sample_values(self, qubit_device_2_wires, tol):
...
         qubit_device_2_wires.apply('RX', wires=[0], par=[1.5708])
 
-        s1 = qubit_device_2_wires.sample('PauliZ', [0], [], 10)
+        s1 = qubit_device_2_wires.sample('PauliZ', [0], [])
 
         # s1 should only contain 1 and -1, which is guaranteed if
         # they square to 1
         assert np.allclose(s1**2, 1, atol=tol, rtol=0)
 
-    def test_sample_exception_analytic_mode(self, qubit_device_2_wires):
```
Comment (co9olguy) [inline tests/test_default_qubit.py]: I get removing this and the following test, but is there an equivalent set of tests for the default.qubit device?

File: pennylane/_device.py [hunk 20]
```diff
@@ -175,6 +174,18 @@ def observables(self):
         """
         raise NotImplementedError
 
+    def set_shots(self, shots):
+        """Changes the number of shots.
+
+        Args:
+            shots (int): number of circuit evaluations/random samples used to estimate
+            expectation values of observables.
```
Comment (co9olguy) [inline pennylane/_device.py]:
```suggestion
            expectation values of observables
```

File: pennylane/_device.py (hunk 7 above)
Comment (co9olguy) [reply pennylane/_device.py]: ``device.shots`` doesn't actually point anywhere. Should it be ``self.shots`` or ``Device.shots``?

File: pennylane/measure.py [hunk 21]
```diff
@@ -148,40 +148,25 @@ def var(op):
     return op
 
 
-def sample(op, n=None):
-    r"""Sample from the supplied observable.
+def sample(op):
+    r"""Sample from the supplied observable with a size of the number of shots
+    given as a parameter to the corresponding Device
```
Comment (co9olguy) [inline pennylane/measure.py]:
```suggestion
    determined from an attribute of the corresponding Device.
```

File: pennylane/measure.py [hunk 22]
```diff
@@ -148,40 +148,25 @@ def var(op):
     return op
 
 
-def sample(op, n=None):
-    r"""Sample from the supplied observable.
+def sample(op):
+    r"""Sample from the supplied observable with a size of the number of shots
```
Comment (co9olguy) [inline pennylane/measure.py]:
```suggestion
    r"""Sample from the supplied observable, with the number of shots
```

File: pennylane/plugins/default_gaussian.py [hunk 23]
```diff
@@ -701,10 +701,12 @@ class DefaultGaussian(Device):
...
     Args:
         wires (int): the number of modes to initialize the device in
-        shots (int): How many times should the circuit be evaluated (or sampled) to estimate
-            the expectation values. 0 yields the exact result.
+        shots (int): (default 1) How many times should the circuit be evaluated (or sampled) to estimate
+            the expectation values.
         hbar (float): (default 2) the value of :math:`\hbar` in the commutation
             relation :math:`[\x,\p]=i\hbar`
+        analytic (bool): (default True) indicates if the device should calculate expectations
+            and variances analytically.
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
            and variances analytically
```

File: pennylane/plugins/default_gaussian.py [hunk 24]
```diff
@@ -701,10 +701,12 @@ class DefaultGaussian(Device):
 
     Args:
         wires (int): the number of modes to initialize the device in
-        shots (int): How many times should the circuit be evaluated (or sampled) to estimate
-            the expectation values. 0 yields the exact result.
+        shots (int): (default 1) How many times should the circuit be evaluated (or sampled) to estimate
+            the expectation values.
```
Comment (co9olguy) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
            the expectation values
```

File: pennylane/_device.py (hunk 5 above)
Comment (johannesjmeyer) [reply pennylane/_device.py]: I actually did `self.shots = shots` because that was what we used before in the `Device.__init__`, but duck typing is fine!

File: pennylane/measure.py (hunk 9 above)
Comment (johannesjmeyer) [reply pennylane/measure.py]: Me too, this is feels much better, too.

File: tests/test_measure.py (hunk 18 above)
Comment (johannesjmeyer) [reply tests/test_measure.py]: Because it tested for an exception that was removed.

File: tests/test_default_qubit.py (hunk 19 above)
Comment (johannesjmeyer) [reply tests/test_default_qubit.py]: I don't understand what you mean, what kind of test do you think should be there?

**Comment by johannesjmeyer** [comment]
I see that 1 is bad, because PennyLane should work "out-of-the-box" and 1 would give very unintuitive behaviour. I have no hard feelings concerning the number of shots used as default. We can do 1000.

File: pennylane/_device.py (hunk 5 above)
Comment (josh146) [reply pennylane/_device.py]: Ah! No worries, sometimes code review can make you critically analyze decisions you forgot you made earlier :laughing:

File: default_config.toml [hunk 25]
```diff
@@ -11,7 +11,7 @@
 ## set the default number of shots/runs used
 ## to determine expectation value statistics
 [main]
-shots = 0
+shots = 1
```
Comment (josh146) [inline default_config.toml]: Should this be 1000 or 1024? Unsure of whether to use a power of 2 or not :laughing:

File: pennylane/_device.py [hunk 26]
```diff
@@ -175,6 +175,26 @@ def observables(self):
         """
         raise NotImplementedError
 
+    @property
+    def shots(self):
+        """Number of circuit evaluations/random samples used to estimate
+            expectation values of observables"""
```
Comment (josh146) [inline pennylane/_device.py]:
```suggestion
        expectation values of observables"""
```

File: pennylane/_device.py [hunk 27]
```diff
@@ -175,6 +175,26 @@ def observables(self):
...
+
+        return self._shots
+
+    @shots.setter
+    def shots(self, shots):
+        """Changes the number of shots.
+
+        Args:
+            shots (int): number of circuit evaluations/random samples used to estimate
+            expectation values of observables
```
Comment (josh146) [inline pennylane/_device.py]:
```suggestion
                expectation values of observables
```

File: pennylane/_device.py [hunk 28]
```diff
@@ -175,6 +175,26 @@ def observables(self):
...
+        """
+        if shots < 1:
+            raise DeviceError("The specified number of shots needs to be at least 1. Got {}.".format(shots))
+
+        self._shots = int(shots)
```
Comment (josh146) [inline pennylane/_device.py]: :+1: I think this is the best approach, because a float will be silently promoted to a integer, whereas anything else should fail with an exception here, rather than failing with a more cryptic error message later.

File: pennylane/configuration.py [hunk 29]
```diff
@@ -56,7 +56,8 @@
     [main]
     # Global PennyLane options.
     # Affects every loaded plugin if applicable.
-    shots = 0
+    shots = 1
```
Comment (josh146) [inline pennylane/configuration.py]: 1000?

File: pennylane/measure.py [hunk 30]
```diff
@@ -148,40 +148,25 @@ def var(op):
     return op
 
 
-def sample(op, n=None):
-    r"""Sample from the supplied observable.
+def sample(op):
+    r"""Sample from the supplied observable, with the number of shots
+    determined from an attribute of the corresponding Device.
```
Comment (josh146) [inline pennylane/measure.py]:
```suggestion
    determined from the ``dev.shots`` attribute of the corresponding device.
```

File: pennylane/plugins/default_gaussian.py [hunk 31]
```diff
@@ -701,10 +701,15 @@ class DefaultGaussian(Device):
 
     Args:
         wires (int): the number of modes to initialize the device in
-        shots (int): How many times should the circuit be evaluated (or sampled) to estimate
-            the expectation values. 0 yields the exact result.
+        shots (int): (default 1000) How many times should the circuit be evaluated (or sampled) to estimate
+            the expectation values.
+            If ``analytic == True``, then the number of shots is ignored
+            in the calculation of expectation values and variances and only controls the number
```
Comment (josh146) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
            in the calculation of expectation values and variances, and only controls the number
```

File: pennylane/plugins/default_gaussian.py [hunk 32]
```diff
@@ -701,10 +701,15 @@ class DefaultGaussian(Device):
 
     Args:
         wires (int): the number of modes to initialize the device in
-        shots (int): How many times should the circuit be evaluated (or sampled) to estimate
-            the expectation values. 0 yields the exact result.
+        shots (int): (default 1000) How many times should the circuit be evaluated (or sampled) to estimate
```
Comment (josh146) [inline pennylane/plugins/default_gaussian.py]:
```suggestion
        shots (int): How many times should the circuit be evaluated (or sampled) to estimate
```

File: pennylane/plugins/default_qubit.py [hunk 33]
```diff
@@ -296,7 +296,12 @@ class DefaultQubit(Device):
     Args:
         wires (int): the number of modes to initialize the device in
         shots (int): How many times the circuit should be evaluated (or sampled) to estimate
-            the expectation values. A value of 0 yields the exact result.
+            the expectation values. Defaults to 1000 if not specified.
+            If ``analytic == True``, then the number of shots is ignored
+            in the calculation of expectation values and variances and only controls the number
```
Comment (josh146) [inline pennylane/plugins/default_qubit.py]:
```suggestion
            in the calculation of expectation values and variances, and only controls the number
```

File: pennylane/_device.py [hunk 34]
```diff
@@ -109,15 +109,15 @@ class Device(abc.ABC):
         wires (int): number of subsystems in the quantum state represented by the device.
             Default 1 if not specified.
         shots (int): number of circuit evaluations/random samples used to estimate
-            expectation values of observables. For simulator devices, a value of 0 results
-            in the exact expectation value being returned. Defaults to 0 if not specified.
+            expectation values of observables. Defaults to 1000 if not specified.
```
Comment (josh146) [inline pennylane/_device.py]: @johannesjmeyer, I think `analytic` should be added to the base Device class, since plugins will also use this new nomenclature. Should be easy to copy and paste it from default_qubit/default_gaussian.

We should also add some logic to disallow analytic=False with shots < 1 here I think, thoughts @co9olguy?

File: tests/test_default_qubit.py (hunk 19 above)
Comment (josh146) [reply tests/test_default_qubit.py]: As a result of this PR, I think this particular edge case could only be achieved via a device such as 
```python
dev = qml.device("default.qubit", wires=2, analytic=False, shots=0)
```
We should add some logic to the base device class to disallow this combination, perhaps.

File: tests/test_measure.py (hunk 18 above)
Comment (co9olguy) [reply tests/test_measure.py]: This particular edge case (trying to sample from a QNode that doesn't have a device attached to it) seems to be independent of the actual PR scope, so I was surprised it was removed (I assumed it was maybe just removed from this file and placed somewhere more appropriate).

If we were checking for this exception before, why stop now? If it is a legitimate edge case that might occur to a user, isn't it better to have a descriptive exception message?

File: tests/test_default_qubit.py (hunk 19 above)
Comment (co9olguy) [reply tests/test_default_qubit.py]: :+1:

File: pennylane/_device.py (hunk 34 above)
Comment (co9olguy) [reply pennylane/_device.py]: Yep

File: default_config.toml (hunk 25 above)
Comment (co9olguy) [inline default_config.toml]: 1000

Comment (co9olguy) [inline default_config.toml]:
```suggestion
shots = 1000
```

File: default_config.toml [hunk 35]
```diff
@@ -46,7 +47,7 @@ hbar = 1
     cutoff_dim = 10
 
     ## set the default number of shots
-    # shots = 0
+    # shots = 1
```
Comment (co9olguy) [inline default_config.toml]:
```suggestion
    # shots = 1000
```

File: default_config.toml [hunk 36]
```diff
@@ -55,7 +56,7 @@ hbar = 1
     ## Default options for the Gaussian simulator
 
     ## set the default number of shots
-    # shots = 0
+    # shots = 1
```
Comment (co9olguy) [inline default_config.toml]:
```suggestion
    # shots = 1000
```

File: pennylane/_device.py [hunk 37]
```diff
@@ -109,15 +109,15 @@ class Device(abc.ABC):
         wires (int): number of subsystems in the quantum state represented by the device.
             Default 1 if not specified.
         shots (int): number of circuit evaluations/random samples used to estimate
```
Comment (co9olguy) [inline pennylane/_device.py]:
```suggestion
        shots (int): Number of circuit evaluations/random samples used to estimate
```

File: pennylane/configuration.py (hunk 29 above)
Comment (co9olguy) [inline pennylane/configuration.py]:
```suggestion
    shots = 1000
```

File: pennylane/configuration.py [hunk 38]
```diff
@@ -90,7 +91,7 @@
 ``[plugin.global]``.
 
 For example, in the above configuration file, the Strawberry Fields
-devices will be loaded with a default of ``shots = 100``, rather than ``shots = 0``. Finally,
+devices will be loaded with a default of ``shots = 100``, rather than ``shots = 1``. Finally,
```
Comment (co9olguy) [inline pennylane/configuration.py]:
```suggestion
devices will be loaded with a default of ``shots = 100``, rather than ``shots = 1000``. Finally,
```

File: pennylane/configuration.py [hunk 39]
```diff
@@ -90,7 +91,7 @@
 ``[plugin.global]``.
```
Comment (co9olguy) [inline pennylane/configuration.py]: Just leaving this as a note here, but should we update projectQ plugin so `num_runs` is replaced by `shots`?

File: tests/test_configuration.py [hunk 40]
```diff
@@ -105,8 +105,8 @@ class TestProperties:
     def test_get_item(self, default_config):
         """Test getting items."""
         # get existing options
-        assert default_config['main.shots'] == 0
-        assert default_config['main']['shots'] == 0
+        assert default_config['main.shots'] == 1
```
Comment (co9olguy) [inline tests/test_configuration.py]:
```suggestion
        assert default_config['main.shots'] == 1000
```

File: tests/test_configuration.py [hunk 41]
```diff
@@ -105,8 +105,8 @@ class TestProperties:
...
+        assert default_config['main']['shots'] == 1
```
Comment (co9olguy) [inline tests/test_configuration.py]:
```suggestion
        assert default_config['main']['shots'] == 1000
```

File: tests/test_configuration.py [hunk 42]
```diff
@@ -158,4 +158,4 @@ def test_device_load(self, default_config):
         dev = qml.device('default.gaussian', wires=2, config=default_config)
 
         assert dev.hbar == 2
-        assert dev.shots == 0
+        assert dev.shots == 1
```
Comment (co9olguy) [inline tests/test_configuration.py]:
```suggestion
        assert dev.shots == 1000
```

File: tests/test_default_gaussian.py [hunk 43]
```diff
@@ -309,7 +309,7 @@ class TestDefaultGaussianDevice(BaseTest):
     """Test the default gaussian device. The test ensures that the device is properly
     applying gaussian operations and calculating the correct observables."""
     def setUp(self):
-        self.dev = DefaultGaussian(wires=2, shots=0, hbar=hbar)
+        self.dev = DefaultGaussian(wires=2, shots=1, hbar=hbar, analytic=True)
```
Comment (co9olguy) [inline tests/test_default_gaussian.py]:
```suggestion
        self.dev = DefaultGaussian(wires=2, shots=1000, hbar=hbar, analytic=True)
```

File: tests/test_default_gaussian.py [hunk 44]
```diff
@@ -419,7 +419,7 @@ def test_apply_errors(self):
 
         with self.assertRaisesRegex(ValueError, "Invalid target subsystems provided in 'wires' argument"):
             p = U2
-            dev = DefaultGaussian(wires=4, shots=0, hbar=hbar)
+            dev = DefaultGaussian(wires=4, shots=1, hbar=hbar)
```
Comment (co9olguy) [inline tests/test_default_gaussian.py]:
```suggestion
            dev = DefaultGaussian(wires=4, shots=1000, hbar=hbar)
```

**Comment by co9olguy** [comment]
From my side, looks almost :100:  now. 
Only two minor things left:
- put back in the exception that tested whether a qnode without a device was trying to access the `shots` parameter
- test the minor edge case that josh mentions

File: tests/test_default_qubit.py (hunk 19 above)
Comment (johannesjmeyer) [reply tests/test_default_qubit.py]: I mean shots=0 is forbidden anyways, so this should not come up, no?

File: pennylane/_device.py (hunk 34 above)
Comment (johannesjmeyer) [reply pennylane/_device.py]: I disagree. `analytic` makes no sense for hardware-only plugins (which are the intended principal target for pennylane), so it should only be in simulator plugins. 

Also, `shots < 1` is already disallowed generally, as it makes no sense for hardware plugins and was replaced by `analytic` for simulator plugins.

Comment (josh146) [reply pennylane/_device.py]: You've changed my mind :slightly_smiling_face:

File: tests/test_measure.py (hunk 18 above)
Comment (johannesjmeyer) [reply tests/test_measure.py]: This particular edge case was there even before sampling was introduced (think of taking an expval of a QNode that has no device attached). The measure routines will yield no problem, because accessing the underlying device is not necessary at this stage. It only caused a problem in `sample` because we pulled the default number of samples from the underlying device (which is not the case anymore, because the number of samples is now a device internal).

I think testing for the current context to have a device attached to it is unnecessary. If we would want to do so, it would have to be in `QNode.__init__` where the device is passed to the `QNode`. We did not need that up to now and I don't think it is important for the future.

File: pennylane/_device.py (hunk 28 above)
Comment (johannesjmeyer) [reply pennylane/_device.py]: :+1:

File: pennylane/_device.py (hunk 34 above)
Comment (johannesjmeyer) [reply pennylane/_device.py]: :+1:

**Comment by johannesjmeyer** [comment]
@co9olguy, see my comments. I think both things you request should not be done or not be necessary.

File: tests/test_default_qubit.py (hunk 19 above)
Comment (co9olguy) [reply tests/test_default_qubit.py]: good point, captured by another test :+1:

File: pennylane/_device.py (hunk 34 above)
Comment (co9olguy) [reply pennylane/_device.py]: Yep, if `shots < 1` is now disallowed, I think we're good without adding more logic.

I have no opinion on the usage of `analytic` though :smile:

**Comment by co9olguy** [comment]
@johannesjmeyer got it, thanks for the discussion! I'm happy now that we're sufficiently covered.
